File: tests/pytest/test_notifications.py
#
#  Copyright (c) 2026 Siddharth Chandrasekaran <sidcha.dev@gmail.com>
#
#  SPDX-License-Identifier: Apache-2.0
#

import os
import time
import threading
import random
import pytest

from osdp import *
from osdp import commands, events
from conftest import make_fifo_pair, cleanup_fifo_pair


PD_ADDR = 101
FILE_ID = 13
FILE_SIZE = 4096


def _fresh_pair(name, enable_notification_on_pd=True):
    key = KeyStore.gen_key()
    f1, f2 = make_fifo_pair(name)

    pd_flags = [LibFlag.EnforceSecure]
    if enable_notification_on_pd:
        pd_flags.append(LibFlag.EnableNotification)
    cp_flags = [LibFlag.EnforceSecure, LibFlag.EnableNotification]

    pd = PeripheralDevice(
        PDInfo(PD_ADDR, f1, scbk=key, flags=pd_flags),
        PDCapabilities([]),
        log_level=LogLevel.Debug,
    )
    cp = ControlPanel(
        [PDInfo(PD_ADDR, f2, scbk=key, flags=cp_flags)],
        log_level=LogLevel.Debug,
    )
    return cp, pd


def _teardown_pair(cp, pd, name):
    try:
        cp.teardown()
    except RuntimeError:
        pass
    try:
        pd.teardown()
    except RuntimeError:
        pass
    cleanup_fifo_pair(name)


class NotifRecorder:
    """Mirrors pd_cmd_cb in tests/unit-tests/test-notifications.c"""

    def __init__(self):
        self.lock = threading.Lock()
        self.pd_status_count = 0
        self.pd_status_last = None
        self.sc_status_count = 0
        self.sc_status_last = None

    def handler(self, cmd):
        if not isinstance(cmd, commands.Notification):
            return None
        with self.lock:
            if cmd.type == NotificationType.PeripheralDeviceStatus:
                self.pd_status_count += 1
                self.pd_status_last = cmd
            elif cmd.type == NotificationType.SecureChannelStatus:
                self.sc_status_count += 1
                self.sc_status_last = cmd
        return None

    def reset_pd_status(self):
        with self.lock:
            self.pd_status_count = 0
            self.pd_status_last = None

    def wait_pd_status(self, timeout):
        return self._wait("pd_status_count", timeout)

    def wait_sc_status(self, timeout):
        return self._wait("sc_status_count", timeout)

    def _wait(self, attr, timeout):
        deadline = time.time() + timeout
        while time.time() < deadline:
            with self.lock:
                if getattr(self, attr) > 0:
                    return True
            time.sleep(0.05)
        return False


def test_pd_receives_pd_status_and_sc_status_notifications():
    """On SC handshake, PD sees PD_STATUS(online=1) and SC_STATUS(active=1)
    via its command callback."""
    cp, pd = _fresh_pair("notif-pd")
    rec = NotifRecorder()
    pd.set_command_handler(rec.handler)
    try:
        pd.start()
        cp.start()
        assert cp.sc_wait_all(timeout=10), "SC handshake did not complete"

        assert rec.wait_pd_status(timeout=5), \
            "PD command callback never saw a PD_STATUS notification"
        assert rec.wait_sc_status(timeout=5), \
            "PD command callback never saw an SC_STATUS notification"

        with rec.lock:
            assert rec.pd_status_last.type == \
                NotificationType.PeripheralDeviceStatus
            assert rec.pd_status_last.arg0 == 1, \
                f"PD_STATUS arg0={rec.pd_status_last.arg0}, want 1 (online)"
            assert rec.sc_status_last.type == \
                NotificationType.SecureChannelStatus
            assert rec.sc_status_last.arg0 == 1, \
                f"SC_STATUS arg0={rec.sc_status_last.arg0}, want 1 (active)"
    finally:
        _teardown_pair(cp, pd, "notif-pd")


def test_pd_receives_offline_notification_on_cp_silence():
    """When CP stops refreshing, PD's online-timeout trips and PD_STATUS
    offline fires on the command callback."""
    cp, pd = _fresh_pair("notif-off")
    rec = NotifRecorder()
    pd.set_command_handler(rec.handler)
    try:
        pd.start()
        cp.start()
        assert cp.sc_wait_all(timeout=10), "SC handshake did not complete"
        assert rec.wait_pd_status(timeout=5), \
            "PD never saw initial online notification"

        rec.reset_pd_status()

        # Silence the CP side. OSDP_PD_ONLINE_TOUT_MS defaults to 8s in
        # shipped builds; give a generous margin for workqueue scheduling.
        cp.stop()

        assert rec.wait_pd_status(timeout=12), \
            "PD never saw offline notification after CP went silent"
        with rec.lock:
            assert rec.pd_status_last.arg0 == 0, \
                f"PD_STATUS arg0={rec.pd_status_last.arg0}, want 0 (offline)"
    finally:
        _teardown_pair(cp, pd, "notif-off")


# -----------------------------------------------------------------------------
# File transfer abort tests
# -----------------------------------------------------------------------------

class _SenderFops:
    def __init__(self, data, tally):
        self._data = data
        self._tally = tally

    def open(self, file_id, file_size):
        assert file_id == FILE_ID
        assert file_size == 0
        return FILE_SIZE

    def read(self, size, offset):
        assert offset < FILE_SIZE
        end = min(offset + size, FILE_SIZE)
        return self._data[offset:end]

    def write(self, data, offset):
        assert False, "sender should not be asked to write"

    def close(self, file_id):
        self._tally["sender_closed"] += 1
        return 0


class _ReceiverFops:
    def __init__(self, received, tally):
        self._received = received
        self._tally = tally

    def open(self, file_id, file_size):
        assert file_id == FILE_ID
        assert file_size == FILE_SIZE
        return 0

    def read(self, size, offset):
        assert False, "receiver should not be asked to read"

    def write(self, data, offset):
        self._received[offset:offset + len(data)] = data
        return len(data)

    def close(self, file_id):
        self._tally["receiver_closed"] += 1
        return 0


def _make_file_ops():
    """Return (sender_fops, receiver_fops, tally) where tally tracks close
    calls so tests can assert the abort path actually closed the backing
    resource.
    """
    data = bytes(random.randint(0, 255) for _ in range(FILE_SIZE))
    received = bytearray(FILE_SIZE)
    tally = {"sender_closed": 0, "receiver_closed": 0}

    return _SenderFops(data, tally), _ReceiverFops(received, tally), tally


def _wait_tx_in_progress(getter, timeout=2.0):
    deadline = time.time() + timeout
    while time.time() < deadline:
        status = getter()
        if status is not None and status.offset > 0:
            return True
        time.sleep(0.02)
    return False


def _wait_tx_cleared(getter, timeout=15.0):
    deadline = time.time() + timeout
    while time.time() < deadline:
        if getter() is None:
            return True
        time.sleep(0.05)
    return False


def _wait_file_tx_done(cp, expected_outcome, timeout=5.0):
    deadline = time.time() + timeout
    while time.time() < deadline:
        remaining = deadline - time.time()
        e = cp.get_event(PD_ADDR, timeout=max(0.05, remaining))
        if e is None:
            continue
        if not isinstance(e, events.Notification):
            continue
        if e.type != NotificationType.FileTransferDone:
            continue
        assert e.file_id == FILE_ID
        assert e.file_tx_outcome == expected_outcome, \
            f"outcome={e.file_tx_outcome}, want {expected_outcome}"
        return e
    pytest.fail(f"FileTransferDone({expected_outcome}) not received")


def test_cp_file_tx_aborts_on_disable_pd():
    """CP side: disabling a PD mid-transfer must abort the file_tx and
    fire FILE_TX_DONE/ABORTED, matching test_cp_file_tx_abort_on_disable
    in the C suite."""
    cp, pd = _fresh_pair("notif-fabort-cp")
    try:
        pd.start()
        cp.start()
        assert cp.sc_wait_all(timeout=10), "SC handshake did not complete"

        sender_fops, receiver_fops, _ = _make_file_ops()
        assert cp.register_file_ops(PD_ADDR, sender_fops)
        assert pd.register_file_ops(receiver_fops)

        # Drain any setup-time notifications.
        while cp.get_event(PD_ADDR, timeout=0) is not None:
            pass

        assert cp.submit_command(PD_ADDR, commands.FileTransfer(id=FILE_ID))

        assert _wait_tx_in_progress(lambda: cp.get_file_tx_status(PD_ADDR)), \
            "file_tx never reached 'in progress' on CP side"

        # Disabling the PD drives OSDP_CP_STATE_DISABLED, which must
        # call osdp_file_tx_abort() and fire FILE_TX_DONE/ABORTED.
        assert cp.disable_pd(PD_ADDR), "disable_pd failed"

        _wait_file_tx_done(cp, FileTxOutcome.Aborted, timeout=5.0)
        assert cp.get_file_tx_status(PD_ADDR) is None, \
            "CP still reports transfer active after abort"
    finally:
        _teardown_pair(cp, pd, "notif-fabort-cp")


def test_pd_file_tx_aborts_on_cp_silence():
    """PD side: when the CP goes silent mid-transfer, the PD's online
    timeout must call osdp_file_tx_abort() on its own transfer."""
    cp, pd = _fresh_pair("notif-fabort-pd")
    try:
        pd.start()
        cp.start()
        assert cp.sc_wait_all(timeout=10), "SC handshake did not complete"

        sender_fops, receiver_fops, _ = _make_file_ops()
        assert cp.register_file_ops(PD_ADDR, sender_fops)
        assert pd.register_file_ops(receiver_fops)

        assert cp.submit_command(PD_ADDR, commands.FileTransfer(id=FILE_ID))

        assert _wait_tx_in_progress(lambda: pd.get_file_tx_status()), \
            "file_tx never reached 'in progress' on PD side"

        # Silence the CP — PD should trip OSDP_PD_ONLINE_TOUT_MS (8s in
        # the shipped build) and abort its in-flight rx.
        cp.stop()

        assert _wait_tx_cleared(lambda: pd.get_file_tx_status(), timeout=15.0), \
            "PD still reports file_tx active after CP went silent"
    finally:
        _teardown_pair(cp, pd, "notif-fabort-pd")

File: tests/unit-tests/test-notifications.c
/*
 * Copyright (c) 2026 Siddharth Chandrasekaran <sidcha.dev@gmail.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <fcntl.h>
#include <pthread.h>

#include <osdp.h>
#include "test.h"

#define NOTIF_FILE_SEND "test-notif-send.txt"
#define NOTIF_FILE_RECV "test-notif-recv.txt"
#define NOTIF_FILE_REPS 200
#define NOTIF_FILE_CHUNK "0123456789abcde\n"
#define NOTIF_FILE_CHUNK_LEN 16

struct notif_record {
	int count;
	int last_type;
	int last_arg0;
	int last_arg1;
};

struct notif_ctx {
	struct test *t;
	osdp_t *cp_ctx;
	osdp_t *pd_ctx;
	int cp_runner;
	int pd_runner;

	pthread_mutex_t lock;
	struct notif_record pd_pd_status;
	struct notif_record pd_sc_status;
	struct notif_record cp_file_tx_done;
	struct notif_record pd_file_tx_done;
	struct notif_record pd_mp_start; /* arg0 = mp.total */
	bool pd_done_before_start;

	struct {
		bool is_cp;
		int fd;
	} sender;
	struct {
		bool is_cp;
		int fd;
	} receiver;
};

static struct notif_ctx g_nx;

static void notif_record_update(struct notif_record *r, int type, int a0, int a1)
{
	r->count++;
	r->last_type = type;
	r->last_arg0 = a0;
	r->last_arg1 = a1;
}

static int pd_cmd_cb(void *arg, struct osdp_cmd *cmd)
{
	struct notif_ctx *nx = arg;

	if (cmd->id != OSDP_CMD_NOTIFICATION) {
		return 0;
	}
	pthread_mutex_lock(&nx->lock);
	switch (cmd->notif.type) {
	case OSDP_NOTIFICATION_PD_STATUS:
		notif_record_update(&nx->pd_pd_status, cmd->notif.type,
				    cmd->notif.pd_status.online, 0);
		break;
	case OSDP_NOTIFICATION_SC_STATUS:
		notif_record_update(&nx->pd_sc_status, cmd->notif.type,
				    cmd->notif.sc_status.active,
				    cmd->notif.sc_status.scbk_d);
		break;
	case OSDP_NOTIFICATION_MP_START:
		notif_record_update(&nx->pd_mp_start, cmd->notif.type,
				    (int)cmd->notif.mp.total,
				    cmd->notif.mp.object_id);
		break;
	case OSDP_NOTIFICATION_MP_DONE:
		if (nx->pd_mp_start.count == 0) {
			nx->pd_done_before_start = true;
		}
		notif_record_update(&nx->pd_file_tx_done, cmd->notif.type,
				    cmd->notif.mp.object_id,
				    cmd->notif.mp.outcome);
		break;
	default: break;
	}
	pthread_mutex_unlock(&nx->lock);
	return 0;
}

static int cp_event_cb(void *arg, int pd, struct osdp_event *ev)
{
	struct notif_ctx *nx = arg;

	ARG_UNUSED(pd);

	if (ev->type != OSDP_EVENT_NOTIFICATION) {
		return 0;
	}
	pthread_mutex_lock(&nx->lock);
	if (ev->notif.type == OSDP_NOTIFICATION_MP_DONE) {
		notif_record_update(&nx->cp_file_tx_done, ev->notif.type,
				    ev->notif.mp.object_id, ev->notif.mp.outcome);
	}
	pthread_mutex_unlock(&nx->lock);
	return 0;
}

static int notif_fops_open(void *arg, int file_id, uint32_t *size)
{
	struct { bool is_cp; int fd; } *t = arg;

	if (file_id != 1 || t->fd != 0)
		return -1;

	t->fd = t->is_cp ? open(NOTIF_FILE_SEND, O_RDONLY)
			 : open(NOTIF_FILE_RECV,
				O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
	if (t->fd < 0) return -1;
	*size = NOTIF_FILE_REPS * NOTIF_FILE_CHUNK_LEN;
	return 0;
}

static int notif_fops_read(void *arg, void *buf, uint32_t size, uint32_t offset)
{
	struct { bool is_cp; int fd; } *t = arg;

	if (t->fd <= 0) return -1;
	return (int)pread(t->fd, buf, (size_t)size, (size_t)offset);
}

static int notif_fops_write(void *arg, const void *buf, uint32_t size, uint32_t offset)
{
	struct { bool is_cp; int fd; } *t = arg;

	if (t->fd <= 0) return -1;
	return (int)pwrite(t->fd, buf, (size_t)size, (size_t)offset);
}

static int notif_fops_close(void *arg)
{
	struct { bool is_cp; int fd; } *t = arg;

	if (t->fd == 0) return -1;
	close(t->fd);
	t->fd = 0;
	return 0;
}

static int notif_create_file(void)
{
	int fd, i;

	fd = open(NOTIF_FILE_SEND, O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
	if (fd < 0) return -1;
	for (i = 0; i < NOTIF_FILE_REPS; i++) {
		if (write(fd, NOTIF_FILE_CHUNK, NOTIF_FILE_CHUNK_LEN) !=
		    NOTIF_FILE_CHUNK_LEN) {
			close(fd);
			return -1;
		}
	}
	close(fd);
	return 0;
}

static bool wait_for_pd_online(int timeout_sec)
{
	int rc = 0;
	uint8_t status = 0;

	while (rc < timeout_sec * 10) {
		osdp_get_status_mask(g_nx.cp_ctx, &status);
		if (status & 1)
			return true;
		usleep(100 * 1000);
		rc++;
	}
	return false;
}

static bool wait_for_notif(struct notif_record *r, int timeout_ms)
{
	int elapsed = 0;
	bool hit;

	while (elapsed <= timeout_ms) {
		pthread_mutex_lock(&g_nx.lock);
		hit = r->count > 0;
		pthread_mutex_unlock(&g_nx.lock);
		if (hit) return true;
		usleep(50 * 1000);
		elapsed += 50;
	}
	return false;
}

static int setup_env(void)
{
	pthread_mutex_init(&g_nx.lock, NULL);

	if (test_setup_devices_ext(g_nx.t, &g_nx.cp_ctx, &g_nx.pd_ctx,
				   OSDP_FLAG_ENABLE_NOTIFICATION,
				   OSDP_FLAG_ENABLE_NOTIFICATION))
		return -1;

	osdp_pd_set_command_callback(g_nx.pd_ctx, pd_cmd_cb, &g_nx);
	osdp_cp_set_event_callback(g_nx.cp_ctx, cp_event_cb, &g_nx);

	g_nx.cp_runner = async_runner_start(g_nx.cp_ctx, osdp_cp_refresh);
	g_nx.pd_runner = async_runner_start(g_nx.pd_ctx, osdp_pd_refresh);
	if (g_nx.cp_runner < 0 || g_nx.pd_runner < 0)
		return -1;

	if (!wait_for_pd_online(10)) {
		printf(SUB_1 "PD did not come online\n");
		return -1;
	}
	return 0;
}

static void teardown_env(void)
{
	struct test *t = g_nx.t;

	if (g_nx.cp_runner >= 0) async_runner_stop(g_nx.cp_runner);
	if (g_nx.pd_runner >= 0) async_runner_stop(g_nx.pd_runner);
	if (g_nx.cp_ctx) osdp_cp_teardown(g_nx.cp_ctx);
	if (g_nx.pd_ctx) osdp_pd_teardown(g_nx.pd_ctx);
	pthread_mutex_destroy(&g_nx.lock);
	memset(&g_nx, 0, sizeof(g_nx));
	g_nx.cp_runner = g_nx.pd_runner = -1;
	g_nx.t = t;
}

static bool test_pd_online_and_sc_notifications(void)
{
	printf(SUB_2 "PD command callback sees PD_STATUS online and SC_STATUS active\n");

	/* Both notifications should already have fired during the
	 * setup_env handshake — setup blocks until the link is up. */
	if (g_nx.pd_pd_status.count == 0) {
		printf(SUB_2 "no PD_STATUS notification on PD side\n");
		return false;
	}
	if (g_nx.pd_pd_status.last_arg0 != 1) {
		printf(SUB_2 "PD_STATUS arg0=%d, want 1 (online)\n",
		       g_nx.pd_pd_status.last_arg0);
		return false;
	}
	if (g_nx.pd_sc_status.count == 0) {
		printf(SUB_2 "no SC_STATUS notification on PD side\n");
		return false;
	}
	if (g_nx.pd_sc_status.last_arg0 != 1) {
		printf(SUB_2 "SC_STATUS arg0=%d, want 1 (active)\n",
		       g_nx.pd_sc_status.last_arg0);
		return false;
	}
	return true;
}

static bool test_pd_offline_on_cp_silence(void)
{
	int deadline_ms;

	printf(SUB_2 "PD transitions offline when CP stops talking\n");

	/* Reset PD_STATUS record so we only observe the offline edge. */
	pthread_mutex_lock(&g_nx.lock);
	memset(&g_nx.pd_pd_status, 0, sizeof(g_nx.pd_pd_status));
	pthread_mutex_unlock(&g_nx.lock);

	/* Silence the CP side. PD should trip the 1.5s timeout (the test
	 * build overrides OSDP_PD_ONLINE_TOUT_MS) and post an offline
	 * notification through its command callback. */
	async_runner_stop(g_nx.cp_runner);
	g_nx.cp_runner = -1;

	/* 1500ms timeout + a 2500ms window for the PD refresh to run
	 * and the workqueue to schedule. */
	deadline_ms = 4000;
	if (!wait_for_notif(&g_nx.pd_pd_status, deadline_ms)) {
		printf(SUB_2 "no offline notification within %dms\n",
		       deadline_ms);
		return false;
	}
	if (g_nx.pd_pd_status.last_arg0 != 0) {
		printf(SUB_2 "PD_STATUS arg0=%d, want 0 (offline)\n",
		       g_nx.pd_pd_status.last_arg0);
		return false;
	}
	return true;
}

static bool test_cp_file_tx_abort_on_disable(void)
{
	struct osdp_cmd cmd = {
		.id = OSDP_CMD_FILE_TX,
		.file_tx = { .id = 1, .flags = 0 },
	};
	struct osdp_file_ops send_ops = {
		.arg = &g_nx.sender,
		.open = notif_fops_open, .read = notif_fops_read,
		.write = notif_fops_write, .close = notif_fops_close,
	};
	struct osdp_file_ops recv_ops = {
		.arg = &g_nx.receiver,
		.open = notif_fops_open, .read = notif_fops_read,
		.write = notif_fops_write, .close = notif_fops_close,
	};
	int rc;
	uint32_t size, offset;

	printf(SUB_2 "CP aborts in-flight transfer when PD is disabled\n");

	g_nx.sender.is_cp = true;
	g_nx.receiver.is_cp = false;

	if (setup_env() != 0) {
		printf(SUB_2 "file-tx: failed to setup env\n");
		return false;
	}
	if (notif_create_file()) {
		printf(SUB_2 "file-tx: failed to create source file\n");
		teardown_env();
		return false;
	}
	if (osdp_file_register_ops(g_nx.cp_ctx, 0, &send_ops) ||
	    osdp_file_register_ops(g_nx.pd_ctx, 0, &recv_ops)) {
		printf(SUB_2 "file-tx: register ops failed\n");
		goto err;
	}
	if (osdp_cp_submit_command(g_nx.cp_ctx, 0, &cmd)) {
		printf(SUB_2 "file-tx: submit failed\n");
		goto err;
	}

	/* Wait until transfer is observably in progress before we yank it. */
	rc = 0;
	while (rc++ < 50) {
		if (osdp_get_file_tx_status(g_nx.cp_ctx, 0, &size, &offset) == 0 &&
		    offset > 0)
			break;
		usleep(20 * 1000);
	}
	if (rc >= 50) {
		printf(SUB_2 "file-tx: never observed tx in progress\n");
		goto err;
	}

	/* Disabling the PD drives OSDP_CP_STATE_DISABLED, which must now
	 * call osdp_file_tx_abort() and fire MP_DONE/ABORTED. */
	pthread_mutex_lock(&g_nx.lock);
	memset(&g_nx.cp_file_tx_done, 0, sizeof(g_nx.cp_file_tx_done));
	pthread_mutex_unlock(&g_nx.lock);

	if (osdp_cp_disable_pd(g_nx.cp_ctx, 0)) {
		printf(SUB_2 "file-tx: disable_pd failed\n");
		goto err;
	}

	if (!wait_for_notif(&g_nx.cp_file_tx_done, 3000)) {
		printf(SUB_2 "file-tx: MP_DONE not seen after disable\n");
		goto err;
	}
	if (g_nx.cp_file_tx_done.last_arg1 != OSDP_FILE_TX_OUTCOME_ABORTED) {
		printf(SUB_2 "file-tx: outcome=%d, want ABORTED=%d\n",
		       g_nx.cp_file_tx_done.last_arg1,
		       OSDP_FILE_TX_OUTCOME_ABORTED);
		goto err;
	}
	if (osdp_get_file_tx_status(g_nx.cp_ctx, 0, &size, &offset) != -1) {
		printf(SUB_2 "file-tx: CP still reports transfer active\n");
		goto err;
	}
	teardown_env();
	unlink(NOTIF_FILE_SEND);
	unlink(NOTIF_FILE_RECV);
	return true;
err:
	teardown_env();
	unlink(NOTIF_FILE_SEND);
	unlink(NOTIF_FILE_RECV);
	return false;
}

static bool test_pd_file_tx_abort_on_offline(void)
{
	struct osdp_cmd cmd = {
		.id = OSDP_CMD_FILE_TX,
		.file_tx = { .id = 1, .flags = 0 },
	};
	struct osdp_file_ops send_ops = {
		.arg = &g_nx.sender,
		.open = notif_fops_open, .read = notif_fops_read,
		.write = notif_fops_write, .close = notif_fops_close,
	};
	struct osdp_file_ops recv_ops = {
		.arg = &g_nx.receiver,
		.open = notif_fops_open, .read = notif_fops_read,
		.write = notif_fops_write, .close = notif_fops_close,
	};
	int rc, elapsed;
	uint32_t size, offset;

	printf(SUB_2 "PD aborts in-flight transfer when CP goes silent\n");

	g_nx.sender.is_cp = true;
	g_nx.receiver.is_cp = false;

	if (setup_env() != 0) {
		printf(SUB_2 "pd-file-tx: failed to setup env\n");
		return false;
	}
	if (notif_create_file()) {
		printf(SUB_2 "pd-file-tx: create file failed\n");
		teardown_env();
		return false;
	}
	if (osdp_file_register_ops(g_nx.cp_ctx, 0, &send_ops) ||
	    osdp_file_register_ops(g_nx.pd_ctx, 0, &recv_ops)) {
		printf(SUB_2 "pd-file-tx: register ops failed\n");
		goto err;
	}
	if (osdp_cp_submit_command(g_nx.cp_ctx, 0, &cmd)) {
		printf(SUB_2 "pd-file-tx: submit failed\n");
		goto err;
	}

	/* Wait until the PD has actually started receiving. */
	rc = 0;
	while (rc++ < 50) {
		if (osdp_get_file_tx_status(g_nx.pd_ctx, 0, &size, &offset) == 0 &&
		    offset > 0)
			break;
		usleep(20 * 1000);
	}
	if (rc >= 50) {
		printf(SUB_2 "pd-file-tx: never observed PD-side rx\n");
		goto err;
	}

	async_runner_stop(g_nx.cp_runner);
	g_nx.cp_runner = -1;

	/* Wait for the PD to trip its online timeout; it should then call
	 * osdp_file_tx_abort() on its own transfer. */
	elapsed = 0;
	while (elapsed < 5000) {
		if (osdp_get_file_tx_status(g_nx.pd_ctx, 0, &size, &offset) == -1)
			break;
		usleep(50 * 1000);
		elapsed += 50;
	}
	if (osdp_get_file_tx_status(g_nx.pd_ctx, 0, &size, &offset) != -1) {
		printf(SUB_2 "pd-file-tx: PD still reports transfer active\n");
		goto err;
	}
	teardown_env();
	unlink(NOTIF_FILE_SEND);
	unlink(NOTIF_FILE_RECV);
	return true;
err:
	teardown_env();
	unlink(NOTIF_FILE_SEND);
	unlink(NOTIF_FILE_RECV);
	return false;
}

/* A clean transfer must deliver the terminal MP_DONE to BOTH roles: the CP
 * (sender) via its event callback and the PD (receiver) via its command
 * callback (synthesized OSDP_CMD_NOTIFICATION). */
static bool test_file_tx_notifies_both_roles(void)
{
	struct osdp_cmd cmd = {
		.id = OSDP_CMD_FILE_TX,
		.file_tx = { .id = 1, .flags = 0 },
	};
	struct osdp_file_ops send_ops = {
		.arg = &g_nx.sender,
		.open = notif_fops_open, .read = notif_fops_read,
		.write = notif_fops_write, .close = notif_fops_close,
	};
	struct osdp_file_ops recv_ops = {
		.arg = &g_nx.receiver,
		.open = notif_fops_open, .read = notif_fops_read,
		.write = notif_fops_write, .close = notif_fops_close,
	};

	printf(SUB_2 "clean transfer notifies both CP and PD\n");

	g_nx.sender.is_cp = true;
	g_nx.receiver.is_cp = false;

	if (setup_env() != 0) {
		printf(SUB_2 "both-roles: failed to setup env\n");
		return false;
	}
	if (notif_create_file()) {
		printf(SUB_2 "both-roles: failed to create source file\n");
		teardown_env();
		return false;
	}
	if (osdp_file_register_ops(g_nx.cp_ctx, 0, &send_ops) ||
	    osdp_file_register_ops(g_nx.pd_ctx, 0, &recv_ops)) {
		printf(SUB_2 "both-roles: register ops failed\n");
		goto err;
	}

	pthread_mutex_lock(&g_nx.lock);
	memset(&g_nx.cp_file_tx_done, 0, sizeof(g_nx.cp_file_tx_done));
	memset(&g_nx.pd_file_tx_done, 0, sizeof(g_nx.pd_file_tx_done));
	memset(&g_nx.pd_mp_start, 0, sizeof(g_nx.pd_mp_start));
	g_nx.pd_done_before_start = false;
	pthread_mutex_unlock(&g_nx.lock);

	if (osdp_cp_submit_command(g_nx.cp_ctx, 0, &cmd)) {
		printf(SUB_2 "both-roles: submit failed\n");
		goto err;
	}

	if (!wait_for_notif(&g_nx.cp_file_tx_done, 5000)) {
		printf(SUB_2 "both-roles: CP never saw MP_DONE\n");
		goto err;
	}
	if (!wait_for_notif(&g_nx.pd_file_tx_done, 5000)) {
		printf(SUB_2 "both-roles: PD never saw MP_DONE\n");
		goto err;
	}
	if (g_nx.cp_file_tx_done.last_arg1 != OSDP_FILE_TX_OUTCOME_OK ||
	    g_nx.pd_file_tx_done.last_arg1 != OSDP_FILE_TX_OUTCOME_OK) {
		printf(SUB_2 "both-roles: outcome cp=%d pd=%d, want OK=%d\n",
		       g_nx.cp_file_tx_done.last_arg1,
		       g_nx.pd_file_tx_done.last_arg1, OSDP_FILE_TX_OUTCOME_OK);
		goto err;
	}
	if (g_nx.cp_file_tx_done.last_arg0 != 1 ||
	    g_nx.pd_file_tx_done.last_arg0 != 1) {
		printf(SUB_2 "both-roles: file_id cp=%d pd=%d, want 1\n",
		       g_nx.cp_file_tx_done.last_arg0,
		       g_nx.pd_file_tx_done.last_arg0);
		goto err;
	}
	/* PD receiver lifecycle: exactly one MP_START, before MP_DONE, and
	 * carrying the declared file size (peeked from the first frame). */
	if (g_nx.pd_mp_start.count != 1 || g_nx.pd_done_before_start) {
		printf(SUB_2 "both-roles: PD start count=%d done-first=%d\n",
		       g_nx.pd_mp_start.count, g_nx.pd_done_before_start);
		goto err;
	}
	if (g_nx.pd_mp_start.last_arg0 !=
	    NOTIF_FILE_REPS * NOTIF_FILE_CHUNK_LEN) {
		printf(SUB_2 "both-roles: PD START total=%d, want %d\n",
		       g_nx.pd_mp_start.last_arg0,
		       NOTIF_FILE_REPS * NOTIF_FILE_CHUNK_LEN);
		goto err;
	}
	teardown_env();
	unlink(NOTIF_FILE_SEND);
	unlink(NOTIF_FILE_RECV);
	return true;
err:
	teardown_env();
	unlink(NOTIF_FILE_SEND);
	unlink(NOTIF_FILE_RECV);
	return false;
}

void run_notification_tests(struct test *t)
{
	printf("\nBegin Notification Tests\n");

	g_nx.t = t;
	g_nx.cp_runner = g_nx.pd_runner = -1;

	/* Group A: PD-side command-callback notifications */
	if (setup_env() != 0) {
		printf(SUB_1 "Failed to setup test environment\n");
		TEST_REPORT(t, false);
		return;
	}
	TEST_CASE(t, "pd_online_and_sc_notifications",
		  test_pd_online_and_sc_notifications());
	TEST_CASE(t, "pd_offline_on_cp_silence",
		  test_pd_offline_on_cp_silence());
	teardown_env();

	/* Group B: file_tx notifications (separate envs so each test
	 * starts from a freshly-online link) */
	TEST_CASE(t, "file_tx_notifies_both_roles",
		  test_file_tx_notifies_both_roles());
	TEST_CASE(t, "cp_file_tx_abort_on_disable",
		  test_cp_file_tx_abort_on_disable());
	TEST_CASE(t, "pd_file_tx_abort_on_offline",
		  test_pd_file_tx_abort_on_offline());
}
